tiny_dnn/layers/max_unpooling_layer.h

```diff
@@ -62,78 +60,77 @@ class max_unpooling_layer : public feedforward_layer<Activation> {
 
-  void back_propagation(serial_size_t index,
-                        const std::vector<vec_t *> &in_data,
-                        const std::vector<vec_t *> &out_data,
-                        std::vector<vec_t *> &out_grad,
-                        std::vector<vec_t *> &in_grad) override {
-    vec_t &prev_delta = *in_grad[0];
-    vec_t &curr_delta = *out_grad[1];
-    std::vector<serial_size_t> &max_idx =
-      max_unpooling_layer_worker_storage_[index].in2outmax_;
+      std::vector<serial_size_t> &max_idx =
+        max_unpooling_layer_worker_storage_.in2outmax_;
 
-    CNN_UNREFERENCED_PARAMETER(in_data);
-
-    this->backward_activation(*out_grad[0], *out_data[0], curr_delta);
+      for_(parallelize_, 0, in2out_.size(), [&](const blocked_range &r) {
+        for (int i = r.begin(); i < r.end(); i++) {
+          const auto &in_index = out2in_[i];
+          out_vec[i] = (max_idx[in_index] == i) ? in_vec[in_index] : float_t{0};
+        }
+      });
+    }
+  }
 
-    for_(parallelize_, 0, in2out_.size(), [&](const blocked_range &r) {
-      for (int i = r.begin(); i != r.end(); i++) {
-        serial_size_t outi = out2in_[i];
-        prev_delta[i] = (max_idx[outi] == i) ? curr_delta[outi] : float_t{0};
-      }
-    });
+  void back_propagation(const std::vector<tensor_t *> &in_data,
+                        const std::vector<tensor_t *> &out_data,
+                        std::vector<tensor_t *> &out_grad,
+                        std::vector<tensor_t *> &in_grad) override {
+    tensor_t &prev_delta = *in_grad[0];
+    tensor_t &curr_delta = *out_grad[0];
+
+    for (serial_size_t sample = 0; sample < in_data[0]->size(); sample++) {
+      vec_t &prev_delta_vec = prev_delta[sample];
+      vec_t &curr_delta_vec = curr_delta[sample];
+
+      std::vector<serial_size_t> &max_idx =
+        max_unpooling_layer_worker_storage_.in2outmax_;
+
+      for_(parallelize_, 0, in2out_.size(), [&](const blocked_range &r) {
+        for (int i = r.begin(); i != r.end(); i++) {
+          serial_size_t outi = out2in_[i];
+          prev_delta_vec[i] =
+            (max_idx[outi] == i) ? curr_delta_vec[outi] : float_t{0};
+        }
+      });
+    }
   }
 
   std::vector<index3d<serial_size_t>> in_shape() const override {
     return {in_};
   }
   std::vector<index3d<serial_size_t>> out_shape() const override {
-    return {out_, out_};
+    return {out_};
   }
   std::string layer_type() const override { return "max-unpool"; }
   size_t unpool_size() const { return unpool_size_; }
 
-  void set_worker_count(serial_size_t worker_count) override {
-    Base::set_worker_count(worker_count);
-    max_unpooling_layer_worker_storage_.resize(worker_count);
-    for (max_unpooling_layer_worker_specific_storage &mws :
-         max_unpooling_layer_worker_storage_) {
-      mws.in2outmax_.resize(out_.size());
-    }
+  void init_workers(serial_size_t sample_count) {
+    max_unpooling_layer_worker_storage_.in2outmax_.resize(out_.size());
   }
```

Comment:
The argument `sample_count` isn't used anymore. What's the point of making this method?
I guess the resizing operation can be simply done in the constructor method.